File: 04_03_Binary_Tree__Trasversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root: return
        lst = []
        queue = collections.deque()
        queue.append(root)
        while queue:
            #size
            n = len(queue)
            
            # create a new sub list for each level
            sublst = []
            
            for _ in range(n):
                
                q = queue.popleft()
                
                sublst.append(q.val)
                if q.left:
                    queue.append(q.left)
                if q.right:
                    queue.append(q.right)
                    
            lst.append(sublst)
            
        return lst
                    

File: 04_03_Binary_Tree_Trasversal.java


// Count row
class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        if (root == null) {
            return new ArrayList<>();
        }
        List<List<Integer>> reVal = new ArrayList<>();
        
        Deque<TreeNode> queue = new ArrayDeque<>();
        
        queue.offer(root);
        
        while(!queue.isEmpty()) {
            
            int current = queue.size();
            List<Integer> row = new ArrayList<>();
            for (int i = 0;i < current;i++) {
                TreeNode currentNode = queue.poll();
                row.add(currentNode.val);

                if (currentNode.left != null) {
                    queue.offer(currentNode.left);
                } 
                if (currentNode.right != null) {
                    queue.offer(currentNode.right);
                }
            }
            reVal.add(row);
        }
        
        System.out.println(reVal);
        
        return reVal;
    }
}


// Two list
class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        if (root == null) {
            return new ArrayList<>();
        }
        List<List<Integer>> reVal = new ArrayList<>();
        
        List<TreeNode> primaryList = new ArrayList<>();
        List<TreeNode> secondaryList = new ArrayList<>();
        
        
        primaryList.add(root);
        
        
        
        while(!primaryList.isEmpty()) {
            List<Integer> row = new ArrayList<>();
            secondaryList = new ArrayList<>();
            
            for(TreeNode currentNode : primaryList) {
                row.add(currentNode.val);
                if (currentNode.left != null) {
                    secondaryList.add(currentNode.left);
                } 
                if (currentNode.right != null) {
                    secondaryList.add(currentNode.right);
                }
            }
            
            primaryList = secondaryList;
            reVal.add(row);
        }
        return reVal;
    }
}
